exchanges/stream/websocket_test.go

```diff
@@ -1448,3 +1445,36 @@ func TestCheckSubscriptions(t *testing.T) {
 	err = ws.checkSubscriptions(nil, subscription.List{{}})
 	assert.NoError(t, err, "checkSubscriptions should not error")
 }
+
+// websocketServerMockEcho is a mock websocket server that echos messages back to the client
+func websocketServerMockEcho(w http.ResponseWriter, r *http.Request) {
+	var upgrader = websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		},
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		fmt.Println("Upgrade error:", err)
+		return
+	}
+	defer conn.Close()
+
+	for {
+		messageType, message, err := conn.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure) {
+				panic(err)
+			}
+			break
+		}
```

Comment:
I think you could use / move `internal/testing/exchange/exchange.go` for most of this.
Your `WsMockFunc` would just do the below.

`internal/testing/websocket/mock.go` might be a good new home to share the mock upgrader.

Reply:
Good suggestion will do